examples/hosts/app-integration/schema-upgrade/src/app.tsx

```diff
@@ -0,0 +1,153 @@
+
+    // Verify the response to make sure we got what we expected.
+    if (response.status !== 200 || response.mimeType !== "fluid/object") {
+        throw new Error(`Unable to retrieve data object at URL: "${url}"`);
+    } else if (response.value === undefined) {
+        throw new Error(`Empty response from URL: "${url}"`);
+    }
+
+    return response.value as IContainerKillBit;
+}
+
+async function start(): Promise<void> {
+    const tinyliciousService = new TinyliciousService();
+
+    const module = { fluidExport: new InventoryListContainerRuntimeFactory() };
+    const codeLoader = { load: async () => module };
+
+    const loader = new Loader({
+        urlResolver: tinyliciousService.urlResolver,
+        documentServiceFactory: tinyliciousService.documentServiceFactory,
+        codeLoader,
+    });
+
+    let fetchedData: string | undefined;
+    let container: IContainer;
+    let inventoryList: IInventoryList;
+    let containerKillBit: IContainerKillBit;
+
+    if (createNew) {
+        fetchedData = await fetchData();
+        container = await loader.createDetachedContainer({ package: "no-dynamic-package", config: {} });
+        inventoryList = await getInventoryListFromContainer(container);
+        containerKillBit = await getContainerKillBitFromContainer(container);
+        await applyStringData(inventoryList, fetchedData);
+        await container.attach({ url: documentId });
+    } else {
+        container = await loader.resolve({ url: documentId });
+        containerKillBit = await getContainerKillBitFromContainer(container);
+        inventoryList = await getInventoryListFromContainer(container);
+    }
+
+    const writeToExternalStorage = async () => {
+        // CONSIDER: it's perhaps more-correct to spawn a new client to extract with (to avoid local changes).
+        // This can be done by making a loader.request() call with appropriate headers (same as we do for the
+        // summarizing client).  E.g.
+        // const exportContainer = await loader.resolve(...);
+        // const inventoryList = (await exportContainer.request(...)).value;
+        // const stringData = extractStringData(inventoryList);
+        // exportContainer.close();
+
+        const stringData = await extractStringData(inventoryList);
+        await writeData(stringData);
+
+        // Normally would be a void, we return the string here for demo purposes only.
+        return stringData;
+    };
+
+    const saveAndEndSession = async () => {
+        if (!containerKillBit.markedForDestruction) {
+            await containerKillBit.markForDestruction();
+        }
+
+        if (containerKillBit.dead) {
+            return;
+        }
+
+        // After the quorum proposal is accepted, our system doesn't allow further edits to the string
+        // So we can immediately get the data out even before taking the lock.
+        const stringData = await extractStringData(inventoryList);
```

Comment:
Possibly multiple clients could end up reading? MarkForDestruction just ensures that op was acked which means that op was broadcasted to other connected clients ?

Reply:
This is true - the Quorum DDS is intended to help fill this gap.